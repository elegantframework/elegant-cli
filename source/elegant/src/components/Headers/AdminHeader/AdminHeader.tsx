import Link from 'next/link';
import { useEffect, useState } from 'react';
import { CMSSignOut } from '@/utils/Auth/hooks';
import docs_light from "@/img/beams/docs-light.png";
import docs_dark from "@/img/beams/docs-dark@tinypng.png";
import clsx from 'clsx';
import Config from '@/utils/Config/Config';
import Logo from '@/components/Logos/Logo/Logo';
import ThemeToggle from '../../Toggles/ThemeToggle/ThemeToggle';

type Props = {
    /**
     * The admin users name.
     */
    name?: string;
    /**
     * The admin users email address.
     */
    email?: string;
    /**
     * The admin users image.
     */
    image?: string;
    /**
     * The admin users status.
     */
    status?: 'authenticated' | 'unauthenticated' | 'loading';
    toggleSidebar: () => void;
    /**
     * Display the background "Beams" image?
     */
    beams?: boolean;
    /**
     * Does this admin header have navigation?
     */
    hasNav?: boolean;
}

/**
 * The admin header to be used within the CMS admin panel.
 * @returns An html admin header element.
 */
export default function AdminHeader({
    name = "",
    email = "",
    image = "",
    status = "unauthenticated",
    toggleSidebar,
    beams = true,
    hasNav = false
}: Props) {

    const [isOpen, setIsOpen] = useState(false);
    const [isOpaque, setIsOpaque] = useState(false);

    useEffect(() => {
        let offset = 50
        function onScroll() {
          if (!isOpaque && window.scrollY > offset) {
            setIsOpaque(true);
          } 
          else if (isOpaque && window.scrollY <= offset) {
            setIsOpaque(false);
          }
        }
        onScroll();
        window.addEventListener('scroll', onScroll, { passive: true });
        return () => {
          window.removeEventListener('scroll', onScroll);
        }
      }, [isOpaque]);

    return <>
        <header
            className={clsx(
            'sticky top-0 z-40 w-full backdrop-blur flex-none transition-colors duration-500 lg:z-50 lg:border-b lg:border-slate-900/10 dark:border-slate-50/[0.06]',
            isOpaque
                ? 'bg-white supports-backdrop-blur:bg-white/95 dark:bg-slate-900/75'
                : 'bg-white/95 supports-backdrop-blur:bg-white/60 dark:bg-transparent'
            )}
        >
            <div className="px-4 sm:px-6 md:px-8">
                {beams && 
                    <div className="absolute z-20 top-0 inset-x-0 flex justify-center overflow-hidden pointer-events-none">
                        <div className="w-[108rem] flex-none flex justify-end">
                            <picture>
                                <img
                                    src={docs_light.src}
                                    alt=""
                                    className="w-[71.75rem] flex-none max-w-none opacity-90 dark:hidden"
                                    decoding="async"
                                />
                            </picture>
                            <picture>
                                <img
                                    src={docs_dark.src}
                                    alt=""
                                    className="w-[90rem] flex-none max-w-none opacity-90 hidden dark:block"
                                    decoding="async"
                                />
                            </picture>
                        </div>
                    </div>
                }
            </div>
            <div className="mx-auto">
                <div
                    className={clsx(
                        'py-4 border-b border-slate-900/10 lg:border-0 dark:border-slate-300/10',
                        hasNav ? 'mx-4 lg:mx-0' : 'px-6'
                    )}
                >
                    <div className="relative flex flex-wrap items-center justify-between">
                        <Link
                            href="/admin"
                            className="mr-3 flex-none w-[2.0625rem] overflow-hidden"
                        >
                            <Logo className="w-auto h-7"/>
                        </Link>
                        {status === 'loading' || (
                            <div className="flex items-center md:order-2">
                                <button
                                    type="button"
                                    className="mr-3 flex items-center rounded-full bg-gray-800 text-sm focus:ring-4 focus:ring-gray-300 dark:focus:ring-gray-600 md:mr-0"
                                    id="user-menu-button"
                                    aria-expanded="false"
                                    data-dropdown-toggle="dropdown"
                                    onClick={() => setIsOpen(!isOpen)}
                                >
                                    <span className="sr-only">Open user menu</span>
                                    {/* eslint-disable-next-line @next/next/no-img-element */}
                                    <img
                                        className="h-8 w-8 rounded-full"
                                        src={image || ''}
                                        alt="user"
                                    />
                                </button>
                                <div
                                    className={`right-0 top-[60px] z-50 my-4 w-full list-none divide-y divide-gray-100 rounded-br rounded-bl bg-white text-base shadow md:-right-0 md:top-[52px] md:w-auto ${
                                        isOpen ? 'block' : 'hidden'
                                    }`}
                                    id="dropdown"
                                    style={{
                                        position: 'absolute',
                                        margin: '0px'
                                    }}
                                >
                                    <div className="py-3 px-4">
                                        <span className="block text-sm text-gray-900">{name}</span>
                                        <span className="block truncate text-sm font-medium text-gray-500">
                                        {email}
                                        </span>
                                    </div>
                                    <ul className="py-1" aria-labelledby="dropdown">
                                        <li>
                                        <a
                                            className="block cursor-pointer py-2 px-4 text-sm text-gray-700 hover:bg-gray-100"
                                            onClick={() => CMSSignOut()}
                                        >
                                            Sign out
                                        </a>
                                        </li>
                                    </ul>
                                </div>
                            </div>
                        )}
                    </div>
                </div>
            </div>
        </header>
    </>;
}