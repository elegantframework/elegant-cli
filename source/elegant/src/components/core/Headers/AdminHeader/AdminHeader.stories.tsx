import type { Meta, StoryObj } from '@storybook/react';
import AdminHeader from './AdminHeader';

const meta: Meta<typeof AdminHeader> = {
  title: 'Core/Headers/Admin Header',
  component: AdminHeader,
};

export default meta;

type Story = StoryObj<typeof AdminHeader>;

export const Default: Story = {
  args: {
  },
};