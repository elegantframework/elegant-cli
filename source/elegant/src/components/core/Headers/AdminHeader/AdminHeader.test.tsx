import React from "react";
import renderer from 'react-test-renderer';
import ReactDOM from "react-dom";
import AdminHeader from "./AdminHeader";

describe("Admin Header component", () => {
    it('renders an admin header properly', () => {
        const header = renderer
        .create(
            <AdminHeader 
                name="Hello World"
                email="test@example.com"
                image=""
                status="authenticated"
                toggleSidebar={() => {}}
            />
        )
        .toJSON();
        expect(header).toMatchSnapshot();
    });
});