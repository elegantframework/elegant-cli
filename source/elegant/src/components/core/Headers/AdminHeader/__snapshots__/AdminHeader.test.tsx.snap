// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Header component renders an admin header properly 1`] = `
[
  <nav
    className="relative border-b border-gray-200 bg-white px-2 py-2.5 sm:px-4 h-14"
  >
    <div
      className="mx-auto flex flex-wrap items-center justify-between"
    >
      <button
        aria-controls="mobile-menu-2"
        aria-expanded="false"
        className="ml-1 inline-flex items-center rounded-lg p-2 text-sm text-gray-500 hover:bg-gray-100 focus:outline-none focus:ring-2 focus:ring-gray-200 dark:text-gray-400 md:hidden"
        data-collapse-toggle="mobile-menu-2"
        onClick={[Function]}
        type="button"
      >
        <span
          className="sr-only"
        >
          Open main menu
        </span>
        <svg
          className="h-6 w-6"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M3 5a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 15a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
            fillRule="evenodd"
          />
        </svg>
        <svg
          className="hidden h-6 w-6"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
            fillRule="evenodd"
          />
        </svg>
      </button>
      <div
        className="cursor-pointer flex items-center"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      />
      <div
        className="flex items-center md:order-2"
      >
        <button
          aria-expanded="false"
          className="mr-3 flex items-center rounded-full bg-gray-800 text-sm focus:ring-4 focus:ring-gray-300 dark:focus:ring-gray-600 md:mr-0"
          data-dropdown-toggle="dropdown"
          id="user-menu-button"
          onClick={[Function]}
          type="button"
        >
          <span
            className="sr-only"
          >
            Open user menu
          </span>
          <img
            alt="user"
            className="h-8 w-8 rounded-full"
            src=""
          />
        </button>
        <div
          className="right-0 top-[60px] z-50 my-4 w-full list-none divide-y divide-gray-100 rounded-br rounded-bl bg-white text-base shadow md:-right-0 md:top-[52px] md:w-auto hidden"
          id="dropdown"
          style={
            {
              "margin": "0px",
              "position": "absolute",
            }
          }
        >
          <div
            className="py-3 px-4"
          >
            <span
              className="block text-sm text-gray-900"
            >
              Hello World
            </span>
            <span
              className="block truncate text-sm font-medium text-gray-500"
            >
              test@example.com
            </span>
          </div>
          <ul
            aria-labelledby="dropdown"
            className="py-1"
          >
            <li>
              <a
                className="block cursor-pointer py-2 px-4 text-sm text-gray-700 hover:bg-gray-100"
                onClick={[Function]}
              >
                Sign out
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </nav>,
  <header
    className="relative mb-10"
  >
    <div
      className="px-4 sm:px-6 md:px-8"
    >
      <div
        className="absolute z-20 top-0 inset-x-0 flex justify-center overflow-hidden pointer-events-none"
      >
        <div
          className="w-[108rem] flex-none flex justify-end"
        >
          <picture>
            <img
              alt=""
              className="w-[71.75rem] flex-none max-w-none dark:hidden"
              decoding="async"
              src="/img.jpg"
            />
          </picture>
          <picture>
            <img
              alt=""
              className="w-[90rem] flex-none max-w-none hidden dark:block"
              decoding="async"
              src="/img.jpg"
            />
          </picture>
        </div>
      </div>
    </div>
  </header>,
]
`;
